src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs

```diff
@@ -85,18 +85,63 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             // No AT in the cache or AT needs to be refreshed
             try
             {
-                using (logger.LogBlockDuration("[OBO request] Fetching OBO token from ESTS"))
-                {
-                    var result = await FetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
-                    return result;
-                }
+                return await RefreshRtOrFetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
             }
             catch (MsalServiceException e)
             {
                 return HandleTokenRefreshError(e, msalAccessTokenItem);
             }
         }
 
+        private async Task<AuthenticationResult> RefreshRtOrFetchNewAccessTokenAsync(CancellationToken cancellationToken)
```

Comment:
if there is a lot code duplication with AcquireTokenSilent, let's try to find a way to reduce it

Reply:
Yes, there are some things different here from the silent flow, in silent flow we fail if the rt cannot be refreshed with UIRequiredException. Here we don't fail. I was also thinking of how to reuse the code. Maybe expose this part of code in some util.